consensus/qbft/core/justification.go

```diff
@@ -0,0 +1,112 @@
+package core
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/consensus/qbft/message"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	istanbul "github.com/ethereum/go-ethereum/consensus/qbft"
+)
+
+// Returns true if the `proposal` is justified by the set `roundChangeMessages` of ROUND-CHANGE messages
+// and by the set `prepareMessages` of PREPARE messages.
+// For this we must either have:
+//     - a quorum of ROUND-CHANGE messages with preparedRound and preparedBlockDigest equal to nil; or
+//     - a ROUND-CHANGE message (1) whose preparedRound is not nil and is equal or higher than the
+//           preparedRound of `quorumSize` ROUND-CHANGE messages and (2) whose preparedRound and
+//           preparedBlockDigest match the round and block of `quorumSize` PREPARE messages.
+func justify(
+	proposal istanbul.Proposal,
+	roundChangeMessages []*message.SignedRoundChangePayload,
+	prepareMessages []*message.SignedPreparePayload,
+	quorumSize int) bool {
+
+	// Check the size of the set of ROUND-CHANGE messages
+	if len(roundChangeMessages) < quorumSize {
+		return false
+	}
+
+	// Check the size of the set of PREPARE messages
+	if len(prepareMessages) != 0 && len(prepareMessages) < quorumSize {
+		return false
+	}
+
+	// If there are PREPARE messages, they all need to have the same round and match `proposal`
+	var preparedRound *big.Int
+	iteration := 0
+	for _, spp := range prepareMessages {
+		if iteration == 0 { // Get the round of the first message
```

Comment:
It seems that the iteration variabl could be avoided by doing something like

```go
var preparedRound *big.Int
if len(prepareMessages) > 0 {
    preparedRound = prepareMessages[0].Round
    for _, spp := range prepareMessages {
        if preparedRound.Cmp(spp.Round) != 0 || proposal.Hash() != spp.Digest {
            return false
        }
    }
}

Reply:
Done